### PR TITLE
modules/managers/import: Fix windows import

### LINKED ISSUE #3285: [Backport] Cobbler Import fails
### Original feature issue

- Issue: #2805
- PRs:
  - #2767
  - #2819

### Target release

- [ ] release33
- [x] release32
- [ ] release30

### Reason

Stabilization of 3.2.x for the community.

### PR DESCRIPTION
**Pull request by SchoolGuy**
## Linked Items

Fixes #3285 

## Description

This PR fixes an issue that was introduced with adding Windows support. It prevented all "cobbler import"s from running successfully.

@tpw56j could you please have a look and review this PR? I am uncertain if I did it right.

## Behaviour changes

Old: "cobbler import" fails

New: "cobbler import" succeedes

## Category

This is related to a:

- [x] Bugfix
- [ ] Feature
- [ ] Packaging
- [ ] Docs
- [ ] Code Quality
- [ ] Refactoring
- [ ] Miscellaneous

## Tests

- [ ] Unit-Tests were created
- [ ] System-Tests were created
- [x] Code is already covered by Unit-Tests
- [ ] Code is already covered by System-Tests
- [ ] No tests required

### CODE REVIEW

File: cobbler/modules/managers/import_signatures.py [hunk 1]
```diff
@@ -240,8 +315,8 @@ def scan_signatures(self):
                                                 break
                                         else:
                                             continue
-                                    self.logger.debug(
-                                        "Found a matching signature: breed=%s, version=%s" % (breed, version))
+                                    self.logger.debug("Found a matching signature: breed=%s, version=%s",
+                                    breed, version)
```
Comment (tpw56j) [inline cobbler/modules/managers/import_signatures.py:319]: This change results in an error:
 ```
received on stderr: 
Exception occurred: <class 'TypeError'>
Exception value: Logger.debug() takes 2 positional arguments but 4 were given
Exception Info:
  File "/usr/lib/python3.11/site-packages/cobbler/remote.py", line 98, in run
    rc = self._run(self)
         ^^^^^^^^^^^^^^^

  File "/usr/lib/python3.11/site-packages/cobbler/remote.py", line 277, in runner
    self.remote.api.import_tree(

  File "/usr/lib/python3.11/site-packages/cobbler/api.py", line 1476, in import_tree
    import_module.run(path, mirror_name, network_root, autoinstall_file, arch, breed, os_version)

  File "/usr/lib/python3.11/site-packages/cobbler/modules/managers/import_signatures.py", line 205, in run
    self.signature = self.scan_signatures()
                     ^^^^^^^^^^^^^^^^^^^^^^

  File "/usr/lib/python3.11/site-packages/cobbler/modules/managers/import_signatures.py", line 318, in scan_signatures
    self.logger.debug("Found a matching signature: breed=%s, version=%s",

!!! TASK FAILED !!!
```

File: cobbler/modules/managers/import_signatures.py [hunk 2]
```diff
@@ -199,6 +213,68 @@ def run(self, path: str, name: str, network_root=None, autoinstall_file=None, ar
...
+            if self.breed == "windows":
+                cmd_path = "/usr/bin/wimexport"
+                bootwim_path = os.path.join(self.path, "sources", "boot.wim")
+                dest_path = os.path.join(self.path, "boot")
+                if os.path.exists(cmd_path) and os.path.exists(bootwim_path):
+                    winpe_path = os.path.join(dest_path, "winpe.wim")
+                    if not os.path.exists(dest_path):
+                        utils.mkdir(dest_path)
+                    rc = utils.subprocess_call([cmd_path, bootwim_path, "1",
+                                                winpe_path, "--boot"], shell=False)
```
Comment (tpw56j) [inline cobbler/modules/managers/import_signatures.py:226]:
```
Exception occurred: <class 'TypeError'>
Exception value: subprocess_call() missing 1 required positional argument: 'cmd'
Exception Info:
  File "/usr/lib/python3.11/site-packages/cobbler/remote.py", line 98, in run
    rc = self._run(self)
         ^^^^^^^^^^^^^^^

  File "/usr/lib/python3.11/site-packages/cobbler/remote.py", line 277, in runner
    self.remote.api.import_tree(

  File "/usr/lib/python3.11/site-packages/cobbler/api.py", line 1476, in import_tree
    import_module.run(path, mirror_name, network_root, autoinstall_file, arch, breed, os_version)

  File "/usr/lib/python3.11/site-packages/cobbler/modules/managers/import_signatures.py", line 225, in run
    rc = utils.subprocess_call([cmd_path, bootwim_path, "1",
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

!!! TASK FAILED !!!
```

**Comment by tpw56j** [comment]
@SchoolGuy Unfortunately, I don't have time to look everything right now, so I'll do it over the weekend.

**Comment by SchoolGuy** [comment]
@tpw56j A day during the next two weeks is enough. Many thanks for taking the time!

File: cobbler/modules/managers/import_signatures.py [hunk 3]
```diff
@@ -199,6 +213,68 @@ def run(self, path: str, name: str, network_root=None, autoinstall_file=None, ar
...
+                    if rc == 0:
+                        cmd = ["/usr/bin/wimdir %s 1 | /usr/bin/grep -i '^/Windows/Boot/PXE$'" % winpe_path]
+                        pxe_path = utils.subprocess_get(cmd, shell=True)[0:-1]
```
Comment (tpw56j) [inline cobbler/modules/managers/import_signatures.py:229]: In version 3.2.2, the utils.subprocess_get method signature looks like this:
```python
def subprocess_call(logger, cmd, shell: bool = True, input=None):
```
`pxe_path = utils.subprocess_get(cmd, shell=True)[0:-1]` -> `pxe_path = utils.subprocess_get(self.logger, cmd, shell=True)[0:-1]`

File: cobbler/modules/managers/import_signatures.py [hunk 4]
```diff
@@ -199,6 +213,68 @@ def run(self, path: str, name: str, network_root=None, autoinstall_file=None, ar
...
+                        cmd = ["/usr/bin/wimdir %s 1 | /usr/bin/grep -i '^/Windows/System32/config/SOFTWARE$'"
+                               % winpe_path]
+                        config_path = utils.subprocess_get(cmd, shell=True)[0:-1]
```
Comment (tpw56j) [inline cobbler/modules/managers/import_signatures.py:232]: `config_path = utils.subprocess_get(cmd, shell=True)[0:-1]` -> 'config_path = utils.subprocess_get(self.logger,  cmd, shell=True)[0:-1]`

File: cobbler/modules/managers/import_signatures.py [hunk 5]
```diff
@@ -199,6 +213,68 @@ def run(self, path: str, name: str, network_root=None, autoinstall_file=None, ar
...
+                        cmd_path = "/usr/bin/wimextract"
+                        rc = utils.subprocess_call([cmd_path, bootwim_path, "1",
+                                                    "%s/pxeboot.n12" % pxe_path,
```
Comment (tpw56j) [inline cobbler/modules/managers/import_signatures.py:235]: `rc = utils.subprocess_call([cmd_path, bootwim_path, "1",` -> `rc = utils.subprocess_call(self.logger, [cmd_path, bootwim_path, "1",`

File: cobbler/modules/managers/import_signatures.py [hunk 6]
```diff
@@ -199,6 +213,68 @@ def run(self, path: str, name: str, network_root=None, autoinstall_file=None, ar
...
+                                    val = h.node_get_value(node, keyname)
+                                    valtype = h.value_type(val)[0]
+                                    value2 = h.value_value(val)[1]
+                                    valobject = {"key": keyname, "t": int(valtype), "value": value2}
+                                    new_values.append(valobject)
+                                    h.node_set_values(node, new_values)
+                                    h.commit(software)
+
+                                    cmd_path = "/usr/bin/wimupdate"
+                                    rc = utils.subprocess_call([cmd_path, winpe_path, "--command=add %s %s"
```
Comment (tpw56j) [inline cobbler/modules/managers/import_signatures.py:270]: `rc = utils.subprocess_call([cmd_path, winpe_path, "--command=add %s %s"` -> 'rc = utils.subprocess_call(self.logger, [cmd_path, winpe_path, "--command=add %s %s"'

File: cobbler/modules/managers/import_signatures.py [hunk 7]
```diff
@@ -199,6 +213,68 @@ def run(self, path: str, name: str, network_root=None, autoinstall_file=None, ar
...
+                                                                % (software, config_path)], shell=False)
+                                    os.remove(software)
```
Comment (tpw56j) [inline cobbler/modules/managers/import_signatures.py:272]: Invalid indentation, this code must be outside the loop `for value in values:`

**Comment by tpw56j** [comment]
The `win.ks` template was not backported, so:
```
Exception occurred: <class 'cobbler.cexceptions.CX'>
Exception value: 'Invalid automatic installation template file location /var/lib/cobbler/templates/win.ks, file not found'
```

**Comment by tpw56j** [comment]
@SchoolGuy  After the above fixes, I was able to successfully import the Windows distro.

**Comment by SchoolGuy** [comment]
@tpw56j I think I caught all your requested changes. I would be very thankful if you could confirm that I did so correctly.

**Comment by tpw56j** [comment]
@SchoolGuy The import ended successfully for me, the windows distro and profile were correctly created.

The purpose of the PR was met, but the result was unusable. To fix this we need to backport `sync_post_wingen.py`, windows code from `tftpgen.py`, `templates/windows/*.template`, `settings`.

**Comment by SchoolGuy** [comment]
@tpw56j Okay got it. I think I need to dive deeper into the Windows feature to backport it completely. For now I need to let this PR rest.

**Comment by tpw56j** [comment]
@SchoolGuy To get deeper into Windows feature, you can start with:
- briefly about the classic Microsoft boot process: https://ipxe.org/appnote/chainload_wds
- alternative to WDS that can be used in conjunction with Cobbler: https://ipxe.org/wimboot

It is not necessary to use `wimboot` for legacy BIOS booting, because `sync_post_wingen.py` can patch the required Microsoft binaries by itself.
But when it became necessary to implement UEFI booting, I found that `wimboot` does the necessary modifications to .efi files on the fly at boot time and there is no need to include such code in `sync_post_wingen.py`.
That is, in today's realities, the simplest option (if you do not know how to patch efi binaries) for deploying Windows using Cobbler relies on `wimboot`.

**Comment by opoplawski** [comment]
Where are we with this?  Do we really need this for a v3.2.3 release or could we let it slip?  Thanks.

**Comment by SchoolGuy** [comment]
@opoplawski I need to time to sit down and do the backport. This is rather important to get Windows working. I find doing this backport rather hard and as such feel free to take it over if you have the resouces.

**Comment by SchoolGuy** [comment]
@tpw56j Is it possible for you to take over this PR? I don't really seem to find the time to finish it and my lack of context for the Windows feature is really not helping.

**Comment by SchoolGuy** [comment]
Thank you for lifting this burden of my shoulder.

**Comment by tpw56j** [comment]
@SchoolGuy  I prepared commit https://github.com/cobbler/cobbler/commit/768ebeca522f3faf2257aa11ed560d3f07061164 for this PR. Could you please move it here. I haven't found how to do this.

**Comment by SchoolGuy** [comment]
@tpw56j As you are not part of the Cobbler Org you have no way to do so sadly (my offer for you still stands). I cherry-picked your commit and added it to this branch.

**Comment by tpw56j** [comment]
@SchoolGuy Maybe it will be easier if I prepare my own PR?

**Comment by SchoolGuy** [comment]
@tpw56j That would be easier yes. Sadly now we have a merge conflict. I believe you are best suited to solve this. I'll close this one then now.

**Pull request closed without merging**